executor/src/witgen/machines/mod.rs

```diff
@@ -151,6 +150,66 @@ impl<'a, T: FieldElement> Machine<'a, T> for KnownMachine<'a, T> {
     }
 }
 
+#[derive(Copy, Clone)]
+/// A reference to a connecting identity, which is either a permutation or a lookup.
+pub enum ConnectingIdentityRef<'a, T> {
+    Lookup(&'a LookupIdentity<T>),
+    Permutation(&'a PermutationIdentity<T>),
+}
+
+impl<'a, T: Display> Display for ConnectingIdentityRef<'a, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ConnectingIdentityRef::Lookup(i) => write!(f, "{i}"),
+            ConnectingIdentityRef::Permutation(i) => write!(f, "{i}"),
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+enum ConnectionType {
+    Lookup,
+    Permutation,
+}
+
+impl<'a, T> ConnectingIdentityRef<'a, T> {
```

Comment:
Is there any way we could re-write the code that uses these Refs so that they are not needed?

Reply:
Hmm.. So assuming we only rely on the overlap of the two (left and right selected expressions) we could just pass that, but I don't really find that cleaner and we lose the ability to do things specifically for each case. 

Also this assumption doesn't seem to hold since we have code like this for `WriteOnceMemory`

```rust
if !parts.connecting_identities.values().all(|i| i.is_lookup()) {
    return None;
}
```

Or this for `BlockMachine`
```
    // Connecting identities should either all be permutations or all lookups.
    let connection_type = connecting_identities
        .values()
        .map(|id| id.kind())
        .unique()
        .exactly_one()
        .ok()?;
```

If your point is that we could refactor these, maybe? I think the current way turns what was a runtime invariant into a compile-time one, with minimal changes to the rest of the code.

Reply:
I think we are not talking about the same thing. My question was if we can avoid adding a `..Ref` type for every type.

Reply:
Ah sorry, I thought this was the code used in path canonicalizer which is also sprinkled with ...Refs.

Reply:
Hm, I still don't particularly like the introduction of this type. Could this also be done using a trait? Then `left` and `right` could be function that are generally useful...

Reply:
I don't really like it either to be honest.

What about

```rust
struct ConnectingIdentity<'a, T> {
   left: &'a SelectedExpressions<T>,
   right: &'a SelectedExpressions<T>,
   kind: ConnectingIdentityKind
}
```

Where `kind` is `Lookup` or `Permutation` and will not need to be extended for `PhantomLookup` and `PhantomPermutation`?

Then we can remove some duplicated code and still rely on `kind` for the cases I pointed out above?
I don't really see how traits help here, since at the end of the day we need to be able to store either of these two possibilities anywhere, so we'd have to use `dyn ConnectingIdentity` which seems unnatural here. But maybe I'm not zooming out enough here.

Reply:
I tried the above and I like it more

Reply:
Cool! Don't we need the ID? Oh, it's in a hash map from id anyway, right?

Reply:
Yeah I put the id for now but it did seem redundant, let's see.

Reply:
It's gone!